File: src/edu/stanford/cfuller/imageanalysistools/filter/LocalBackgroundEstimation3DFilter.java
/* ***** BEGIN LICENSE BLOCK *****
 * 
 * Copyright (c) 2011 Colin J. Fuller
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * 
 * ***** END LICENSE BLOCK ***** */

package edu.stanford.cfuller.imageanalysistools.filter;


import edu.stanford.cfuller.imageanalysistools.image.Image;
import edu.stanford.cfuller.imageanalysistools.image.ImageCoordinate;

/**
 * A Filter that estimates the background locally in an Image, using a local mean filtering approach.  It is intended for use on three-dimensional images.
 * <p>
 * This filter may be useful for determining and correcting for local intensity variations.
 * <p>
 * The reference image should be set to the Image that is to be median filtered.  This Image will not be changed.
 * <p>
 * The argument to the apply method should be any Image (except a shallow copy of the reference Image) of the same dimensions as the reference Image.
 * The mean filtered Image will be written to this Image.
 * <p>
 *
 * @author Colin J. Fuller
 * 
 */
public class LocalBackgroundEstimation3DFilter extends LocalBackgroundEstimationFilter {

	
	public LocalBackgroundEstimation3DFilter() {
	}
	
	/**
     * Applies the LocalBackgroundEstimationFilter to an Image.
     * @param im    The Image that will be replaced by the output Image.  This can be anything of the correct dimensions except a shallow copy of the reference Image.
     */
	@Override
	public void apply(Image im) {

		if (this.referenceImage == null) return;
				
		ImageCoordinate boxMin = ImageCoordinate.cloneCoord(this.referenceImage.getDimensionSizes());
		for (Integer s : boxMin) {
			boxMin.set(s, 0);
		}
		
		ImageCoordinate boxMax = ImageCoordinate.cloneCoord(this.referenceImage.getDimensionSizes());
		
		//boolean first = true;
				
		Image counts = new Image(this.referenceImage.getDimensionSizes(), 0.0);
		
		ImageCoordinate lastCoordinate = ImageCoordinate.createCoordXYZCT(0, 0, 0, 0, 0);
		
		for(ImageCoordinate i : im) {
						
			//if (first) {
			
			if (i.get(ImageCoordinate.X) == 0 ) {
			
				boxMin.set(ImageCoordinate.X, i.get(ImageCoordinate.X) - boxSize);
				boxMin.set(ImageCoordinate.Y, i.get(ImageCoordinate.Y) - boxSize);
				boxMin.set(ImageCoordinate.Z, i.get(ImageCoordinate.Z) - boxSize);
				
				boxMax.set(ImageCoordinate.X, i.get(ImageCoordinate.X) + boxSize+1);
				boxMax.set(ImageCoordinate.Y, i.get(ImageCoordinate.Y) + boxSize+1);
				boxMax.set(ImageCoordinate.Z, i.get(ImageCoordinate.Z) + boxSize+1);
				
				this.referenceImage.setBoxOfInterest(boxMin, boxMax);
			
				double total = 0;
				int count = 0;
				
				for (ImageCoordinate iBox : this.referenceImage) {
					
					total += this.referenceImage.getValue(iBox);
					count++;
					
				}
				
				im.setValue(i, total/count);
				counts.setValue(i, count);
				
				this.referenceImage.clearBoxOfInterest();
				
				//first = false;
								
			} else {
				
				lastCoordinate.setCoord(i);
				
				int x = i.get(ImageCoordinate.X);
				int y = i.get(ImageCoordinate.Y);
				int z = i.get(ImageCoordinate.Z);
								
				double sum = 0;
				
				int count = 0;
				
				if (x > 0) {
										
					lastCoordinate.set(ImageCoordinate.X, x-1);
					
					count = (int) counts.getValue(lastCoordinate);
					
					sum = im.getValue(lastCoordinate) * count;
					
					boxMin.set(ImageCoordinate.X, x - boxSize-1);
					boxMin.set(ImageCoordinate.Y, y - boxSize);
					boxMin.set(ImageCoordinate.Z, z - boxSize);
					
					boxMax.set(ImageCoordinate.X, x-boxSize);
					boxMax.set(ImageCoordinate.Y, i.get(ImageCoordinate.Y) + boxSize+1);
					boxMax.set(ImageCoordinate.Z, i.get(ImageCoordinate.Z) + boxSize+1);
					
					this.referenceImage.setBoxOfInterest(boxMin, boxMax);
										
					for (ImageCoordinate iBox : this.referenceImage) {
												
						sum -= this.referenceImage.getValue(iBox);
						count--;
						
					}
					
					this.referenceImage.clearBoxOfInterest();
					
					boxMin.set(ImageCoordinate.X, x+boxSize);
					boxMax.set(ImageCoordinate.X, x+boxSize+1);
					
					this.referenceImage.setBoxOfInterest(boxMin, boxMax);

					for (ImageCoordinate iBox : this.referenceImage) {
						try {
						sum += this.referenceImage.getValue(iBox);
						count++;
						} catch (ArrayIndexOutOfBoundsException e) {
							System.err.println(iBox);
							System.err.println(this.referenceImage.getDimensionSizes());
							System.err.println(this.referenceImage.getBoxMax());
							System.err.println(this.referenceImage.getBoxMin());
							throw e;
						}
						
					}
					
					im.setValue(i, sum/count);
					counts.setValue(i, count);
					
					this.referenceImage.clearBoxOfInterest();
					
				} else if (y > 0) {
					
					//TODO implement this instead of complete calculation for the first column
					
				} else if (z > 0) {
					
					//TODO implement this instead of complete calculation for the first column

				}
				
				
			}
			
		}
		
		lastCoordinate.recycle();
		boxMin.recycle();
		boxMax.recycle();

		
	}
	
	
}

File: src/main/kotlin/edu/stanford/cfuller/imageanalysistools/filter/LocalBackgroundEstimation3DFilter.kt
package edu.stanford.cfuller.imageanalysistools.filter

import edu.stanford.cfuller.imageanalysistools.image.WritableImage
import edu.stanford.cfuller.imageanalysistools.image.ImageCoordinate
import edu.stanford.cfuller.imageanalysistools.image.ImageFactory

/**
 * A Filter that estimates the background locally in an Image, using a local mean filtering approach.  It is intended for use on three-dimensional images.
 *
 *
 * This filter may be useful for determining and correcting for local intensity variations.
 *
 *
 * The reference image should be set to the Image that is to be median filtered.  This Image will not be changed.
 *
 *
 * The argument to the apply method should be any Image (except a shallow copy of the reference Image) of the same dimensions as the reference Image.
 * The mean filtered Image will be written to this Image.
 *
 *

 * @author Colin J. Fuller
 */
class LocalBackgroundEstimation3DFilter : LocalBackgroundEstimationFilter() {
    /**
     * Applies the LocalBackgroundEstimationFilter to an Image.
     * @param im    The Image that will be replaced by the output Image.  This can be anything of the correct dimensions except a shallow copy of the reference Image.
     */
    override fun apply(im: WritableImage) {
        val referenceImage = this.referenceImage ?:
                throw ReferenceImageRequiredException("LocalBackgroundEstimation3DFilter requires a reference image.")

        val boxMin = ImageCoordinate.cloneCoord(referenceImage.dimensionSizes)
        for (s in boxMin) {
            boxMin[s] = 0
        }

        val boxMax = ImageCoordinate.cloneCoord(referenceImage.dimensionSizes)
        var first = true
        val counts = ImageFactory.createWritable(referenceImage.dimensionSizes, 0.0f)
        val lastCoordinate = ImageCoordinate.createCoordXYZCT(0, 0, 0, 0, 0)

        for (i in im) {
            if (first) {
                first = false
                boxMin[ImageCoordinate.X] = i[ImageCoordinate.X] - boxSize
                boxMin[ImageCoordinate.Y] = i[ImageCoordinate.Y] - boxSize
                boxMin[ImageCoordinate.Z] = i[ImageCoordinate.Z] - boxSize

                boxMax[ImageCoordinate.X] = i[ImageCoordinate.X] + boxSize + 1
                boxMax[ImageCoordinate.Y] = i[ImageCoordinate.Y] + boxSize + 1
                boxMax[ImageCoordinate.Z] = i[ImageCoordinate.Z] + boxSize + 1

                referenceImage.setBoxOfInterest(boxMin, boxMax)

                var total = 0f
                var count = 0

                for (iBox in referenceImage) {
                    total += referenceImage.getValue(iBox)
                    count++
                }

                im.setValue(i, total / count)
                counts.setValue(i, count.toFloat())
                referenceImage.clearBoxOfInterest()
            } else {
                lastCoordinate.setCoord(i)
                val x = i[ImageCoordinate.X]
                val y = i[ImageCoordinate.Y]
                val z = i[ImageCoordinate.Z]
                var sum = 0f
                var count = 0
                if (x > 0) {
                    lastCoordinate[ImageCoordinate.X] = x - 1
                    count = counts.getValue(lastCoordinate).toInt()
                    sum = im.getValue(lastCoordinate) * count

                    boxMin[ImageCoordinate.X] = x - boxSize - 1
                    boxMin[ImageCoordinate.Y] = y - boxSize
                    boxMin[ImageCoordinate.Z] = z - boxSize

                    boxMax[ImageCoordinate.X] = x - boxSize
                    boxMax[ImageCoordinate.Y] = i[ImageCoordinate.Y] + boxSize + 1
                    boxMax[ImageCoordinate.Z] = i[ImageCoordinate.Z] + boxSize + 1

                    referenceImage.setBoxOfInterest(boxMin, boxMax)

                    for (iBox in referenceImage!!) {
                        sum -= referenceImage.getValue(iBox)
                        count--
                    }

                    referenceImage.clearBoxOfInterest()

                    boxMin[ImageCoordinate.X] = x + boxSize
                    boxMax[ImageCoordinate.X] = x + boxSize + 1

                    referenceImage.setBoxOfInterest(boxMin, boxMax)

                    for (iBox in referenceImage) {
                        try {
                            sum += referenceImage.getValue(iBox)
                            count++
                        } catch (e: ArrayIndexOutOfBoundsException) {
                            System.err.println(iBox)
                            System.err.println(referenceImage.dimensionSizes)
                            System.err.println(referenceImage.boxMax)
                            System.err.println(referenceImage.boxMin)
                            throw e
                        }
                    }

                    im.setValue(i, sum / count)
                    counts.setValue(i, count.toFloat())
                    referenceImage.clearBoxOfInterest()
                } else if (y > 0) {
                    lastCoordinate[ImageCoordinate.Y] = y - 1
                    count = counts.getValue(lastCoordinate).toInt()
                    sum = im.getValue(lastCoordinate) * count

                    boxMin[ImageCoordinate.X] = x - boxSize
                    boxMin[ImageCoordinate.Y] = y - boxSize - 1
                    boxMin[ImageCoordinate.Z] = z - boxSize

                    boxMax[ImageCoordinate.X] = i[ImageCoordinate.X] + boxSize + 1
                    boxMax[ImageCoordinate.Y] = y - boxSize
                    boxMax[ImageCoordinate.Z] = i[ImageCoordinate.Z] + boxSize + 1

                    referenceImage.setBoxOfInterest(boxMin, boxMax)

                    for (iBox in referenceImage) {
                        sum -= referenceImage.getValue(iBox)
                        count--
                    }

                    referenceImage.clearBoxOfInterest()

                    boxMin[ImageCoordinate.Y] = y + boxSize
                    boxMax[ImageCoordinate.Y] = y + boxSize + 1

                    referenceImage.setBoxOfInterest(boxMin, boxMax)

                    for (iBox in referenceImage) {
                        try {
                            sum += referenceImage.getValue(iBox)
                            count++
                        } catch (e: ArrayIndexOutOfBoundsException) {
                            System.err.println(iBox)
                            System.err.println(referenceImage.dimensionSizes)
                            System.err.println(referenceImage.boxMax)
                            System.err.println(referenceImage.boxMin)
                            throw e
                        }
                    }

                    im.setValue(i, sum / count)
                    counts.setValue(i, count.toFloat())
                    referenceImage.clearBoxOfInterest()
                } else if (z > 0) {
                    lastCoordinate[ImageCoordinate.Z] = z - 1
                    count = counts.getValue(lastCoordinate).toInt()
                    sum = im.getValue(lastCoordinate) * count

                    boxMin[ImageCoordinate.X] = x - boxSize
                    boxMin[ImageCoordinate.Y] = y - boxSize
                    boxMin[ImageCoordinate.Z] = z - boxSize - 1

                    boxMax[ImageCoordinate.X] = i[ImageCoordinate.X] + boxSize + 1
                    boxMax[ImageCoordinate.Y] = i[ImageCoordinate.Y] + boxSize + 1
                    boxMax[ImageCoordinate.Z] = z - boxSize

                    referenceImage.setBoxOfInterest(boxMin, boxMax)

                    for (iBox in referenceImage) {
                        sum -= referenceImage.getValue(iBox)
                        count--
                    }

                    referenceImage.clearBoxOfInterest()

                    boxMin[ImageCoordinate.Z] = z + boxSize
                    boxMax[ImageCoordinate.Z] = z + boxSize + 1

                    referenceImage.setBoxOfInterest(boxMin, boxMax)

                    for (iBox in referenceImage) {
                        try {
                            sum += referenceImage.getValue(iBox)
                            count++
                        } catch (e: ArrayIndexOutOfBoundsException) {
                            System.err.println(iBox)
                            System.err.println(referenceImage.dimensionSizes)
                            System.err.println(referenceImage.boxMax)
                            System.err.println(referenceImage.boxMin)
                            throw e
                        }
                    }
                    im.setValue(i, sum / count)
                    counts.setValue(i, count.toFloat())
                    referenceImage.clearBoxOfInterest()
                }
            }
        }
        lastCoordinate.recycle()
        boxMin.recycle()
        boxMax.recycle()
    }
}
